platform-sdk/docs/proposals/001-tss-rosters-design-proposal/proposal.md

```diff
@@ -0,0 +1,447 @@
+     * SHALL NOT supply both values for the same endpoint.<br/>
+     * This list SHALL NOT be empty.<br/>
+     * This list SHALL NOT contain more than `10` entries.<br/>
+     * The first two entries in this list SHALL be the endpoints published to
+     * all consensus nodes.<br/>
+     * All other entries SHALL be reserved for future use.
+     */
+    repeated proto.ServiceEndpoint gossip_endpoint = 5;
+}
+```
+
+The map of rosters will be stored in the state as a virtual map of Key `Hash<Roster>` and value `Roster`.
+The `Hash` object(SHA_384), which is the key of the virtual map, will reuse an existing protobuf that looks as follows:
+```proto
+/**
+ * List of hash algorithms
+ */
+enum HashAlgorithm {
+  HASH_ALGORITHM_UNKNOWN = 0;
+  SHA_384 = 1;
+}
+
+/**
+ * Encapsulates an object hash so that additional hash algorithms
+ * can be added in the future without requiring a breaking change.
+ */
+message HashObject {
+
+  /**
+   * Specifies the hashing algorithm
+   */
+  HashAlgorithm algorithm = 1;
+
+  /**
+   * Hash length
+   */
+  int32 length = 2;
+
+  /**
+   * Specifies the result of the hashing operation in bytes
+   */
+  bytes hash = 3;
+}
+```
+
+The `Roster` value is modeled as previously shown.
+
+The platform state protobuf will be updated to include the roster hashes.
+
+```proto
+/**
+ * The current state of platform consensus.<br/>
+ * This message stores the current consensus data for the platform
+ * in network state.
+ *
+ * The platform state SHALL represent the latest round's consensus.<br/>
+ * This data SHALL be used to ensure consistency and provide critical data for
+ * restart and reconnect.
+ */
+message PlatformState {
+
+    //... existing entries omitted for brevity ...
+
+    /**
+     * The SHA-384 hash of a candidate roster.
+     * <p>
+     * This is the hash of the roster that is currently being considered
+     * for adoption.
+     * A Node SHALL NOT, ever, have more than one candidate roster
+     * at the same time.
+     */
+    bytes candidate_roster_hash = 6;
+
+    /**
+     * The SHA-384 hash of an active roster
+     * This is the hash of the roster that has already been adopted by the
+     * network.
+     * A Node SHALL NOT, ever, have more than one active roster
+     * at the same time.
+     */
+    bytes active_roster_hash = 7;
+}
+```
+
+It is noteworthy that the roster must be immutable to guarantee the integrity of the computed hash.
+This map of rosters will typically contain the current Active Roster, an optional previously Active Roster, and an
+optional current Candidate Roster.
+There will be new fields in PlatformState - `candidateRosterHash` and `activeRosterHash` - such that at adoption time,
+the app will set the roster through the API and the platform code inserting this roster into the roster map, alongside
+setting the `candidateRosterHash` field in the PlatformState. If a `candidateRosterHash` hash entry already exist in the
+map of Rosters, it will be discarded.
```

Comment:
>If a `candidateRosterHash` hash entry already exist in the
map of Rosters, it will be discarded.
I think that describes idempotency, but I've included the word specifically. Is that enough?

Reply:
Looks good now. Thanks!